arrayfire/blas.py

```diff
@@ -202,3 +202,109 @@ def dot(lhs, rhs, lhs_opts=MATPROP.NONE, rhs_opts=MATPROP.NONE, return_scalar =
+
+    alpha : scalar
+
+    beta : scalar
+
+    lhs_opts: optional: af.MATPROP. default: af.MATPROP.NONE.
+              Can be one of
+               - af.MATPROP.NONE   - If no op should be done on `lhs`.
+               - af.MATPROP.TRANS  - If `lhs` has to be transposed before multiplying.
+               - af.MATPROP.CTRANS - If `lhs` has to be hermitian transposed before multiplying.
+
+    rhs_opts: optional: af.MATPROP. default: af.MATPROP.NONE.
+              Can be one of
+               - af.MATPROP.NONE   - If no op should be done on `rhs`.
+               - af.MATPROP.TRANS  - If `rhs` has to be transposed before multiplying.
+               - af.MATPROP.CTRANS - If `rhs` has to be hermitian transposed before multiplying.
+
+    Returns
+    -------
+
+    out : af.Array
+          Output of the matrix multiplication on `lhs` and `rhs`.
+
+    Note
+    -----
+
+    - The data types of `lhs` and `rhs` should be the same.
+    - Batches are not supported.
+
+    """
+    if C is None:
+        out = Array()
+    else:
+        out = C
+
+    ltype = lhs.dtype()
+
+    if ltype == Dtype.f32:
+        aptr = c_cast(c_pointer(c_float_t(alpha)),c_void_ptr_t)
+        bptr = c_cast(c_pointer(c_float_t(beta)), c_void_ptr_t)
+    elif ltype == Dtype.c32:
+        if isinstance(alpha, af_cfloat_t):
+            aptr = c_cast(c_pointer(alpha), c_void_ptr_t)
+        elif isinstance(alpha, tuple):
+            aptr = c_cast(c_pointer(af_cfloat_t(alpha[0], alpha[1])), c_void_ptr_t)
+        else:
+            aptr = c_cast(c_pointer(af_cfloat_t(alpha)), c_void_ptr_t)
+
+        if isinstance(beta, af_cfloat_t):
+            bptr = c_cast(c_pointer(beta), c_void_ptr_t)
+        elif isinstance(beta, tuple):
+            bptr = c_cast(c_pointer(af_cfloat_t(beta[0], beta[1])), c_void_ptr_t)
+        else:
+            bptr = c_cast(c_pointer(af_cfloat_t(beta)), c_void_ptr_t)
+
+    elif ltype == Dtype.f64:
+        aptr = c_cast(c_pointer(c_double_t(alpha)),c_void_ptr_t)
+        bptr = c_cast(c_pointer(c_double_t(beta)), c_void_ptr_t)
+    elif ltype == Dtype.c64:
+        if isinstance(alpha, af_cdouble_t):
+            aptr = c_cast(c_pointer(alpha), c_void_ptr_t)
+        elif isinstance(alpha, tuple):
+            aptr = c_cast(c_pointer(af_cdouble_t(alpha[0], alpha[1])), c_void_ptr_t)
+        else:
+            aptr = c_cast(c_pointer(af_cdouble_t(alpha)), c_void_ptr_t)
+
+        if isinstance(beta, af_cdouble_t):
+            bptr = c_cast(c_pointer(beta), c_void_ptr_t)
+        elif isinstance(beta, tuple):
+            bptr = c_cast(c_pointer(af_cdouble_t(beta[0], beta[1])), c_void_ptr_t)
+        else:
+            bptr = c_cast(c_pointer(af_cdouble_t(beta)), c_void_ptr_t)
+    elif ltype == Dtype.f16:
+        raise TypeError("fp16 currently unsupported gemm() input type")
+    else:
+        raise TypeError("unsupported input type")
+
+
+    safe_call(backend.get().af_gemm(c_pointer(out.arr),
```

Comment:
Unrelated to this PR: I am curious if we have to do `backend.get()` for every such call. Is this cost negligible ?